Checkout: Rearrange Terms of Service into a foldable card component

jjchrisdiehl opened the pull request:
Related to https://github.com/Automattic/payments-shilling/issues/1969#issuecomment-1800291137

These changes move _unrequired_ terms of service items into a foldable card to help reduce the wall of text seen in some shopping carts, especially carts with lots of promotions and different products.

**Note**: 'unrequired terms' are terms that are not legally required to be shown in the terms of service, or, are not required by payment providers or third party partners.

Additionally, an `x-day money back guarantee for [product]` has been added below the cart submit button.

Experiment code PR https://github.com/Automattic/wp-calypso/pull/84395

## Proposed Changes

So far have a collapsible ToS section. Still need to work out what terms will go where...

**Collapsed**

![image](https://github.com/Automattic/wp-calypso/assets/16580129/7078f9a4-e0a5-42c5-a286-75a88c382c9b)

**Expanded**

![image](https://github.com/Automattic/wp-calypso/assets/16580129/2bf9461d-e294-4739-bdd0-09dd36979874)

**Guarantee on desktop**

![image](https://github.com/Automattic/wp-calypso/assets/16580129/9ded2887-84b2-4b07-b52e-a90cc30570e1)

**Guarantee on mobile**

![image](https://github.com/Automattic/wp-calypso/assets/16580129/d0c7962e-bbc2-4b9b-8af1-718ff24ffccf)

(this could be centered, thoughts?)

## Testing Instructions
- Add a few products to your cart such as a plan, a custom domain, a professional email etc.
- Go to checkout and add the following query parameter to the URL `?checkoutVersion=tos-foldable-card`, you can also see the changes in `?checkoutVersion=2`.
- Ensure that the ToS section shows the collapsible card depicted above, and that the card opens and closes as expected
- Check that the card and ToS section resize on different screen widths correctly
- Check that the money back guarantee shows correctly, and also shows the correct refund period for the product it highlights
- Make sure that the guarantee on mobile doesn't cover the ToS section
- Lastly, check regular checkout to ensure none of the above has bled into the production checkout

Commit:
Remove icons from ToS items

Commit:
Add FoldableCard to ToS section

JanaMW27 commented:
I would add the chevron beside the sentence Read more about your order's terms of service and make the chevron 20px by 20px, all to make it much more subtle.

Commit:
Restyle foldable card header

Commit:
Add CheckoutMoneyBackGuarantee component

Commit:
Add the CheckoutMoneyBackGuarantee component file

Commit:
Move guarantee component inside CheckoutFormSubmit

jjchrisdiehl commented:
Most of the way there, so far I have a pretty decent approximation of the final product:

**Closed**

![image](https://github.com/Automattic/wp-calypso/assets/16580129/03bc9af9-571f-443a-9d0f-4a1baa6df4ca)

**Open**

![image](https://github.com/Automattic/wp-calypso/assets/16580129/9404c42e-c095-4295-a4a2-68a5364b5035)

Though I noticed on mobile widths that the guarantee wasn't repositioning along with the CheckoutFormSubmit so with my last commit I made some changes to nest the guarantee component within the CheckoutFormSubmit component. It currently works as needed, but I need to add a bit more styling to compensate for the change.

![image](https://github.com/Automattic/wp-calypso/assets/16580129/bfce416f-9a75-42f3-88e9-49b330fe2a51)

Commit:
Update styling of guarantee on mobile

DavidRothstein commented:
Some quick feedback, related to what I mentioned on Slack a few days ago:

>Read more about your order's terms of service

This isn't accurate wording.  You don't see the terms of service when you click on it; the terms of service are a specific legal agreement already linked in the first paragraph.  They shouldn't be mentioned again here.

Perhaps just use "Read more"?

Commit:
Add feature flag

Commit:
Add helper functions for feature flags

jjchrisdiehl commented:
> Perhaps just use "Read more"?

This works for me, it would also help with fitting the headline for this foldable card onto one line with smaller devices.

Commit:
Update foldable card header text

Commit:
Convert helper functions to getQueryParam

Commit:
Remove RefundTerms100Year from current branch

jjchrisdiehl commented:
The 100 Year Plan guarantee is fixed now due to the changes merged in https://github.com/Automattic/wp-calypso/pull/84125 and  https://github.com/Automattic/wp-calypso/pull/84209

![image](https://github.com/Automattic/wp-calypso/assets/16580129/9ca916c1-d358-404f-87c9-42c6d4aa0343)

Commit:
Add useToSFoldableCard hook

Commit:
Update Accept Terms checkbox styles

Commit:
Hide gridicons in tos

Commit:
Reposition NonTotalPrices and restyle

JanaMW27 commented:
Hello, this is my bad, but could we replace the icon on the refund line to use the TRASNFORM icon?

<img width="122" alt="Screenshot 2023-11-16 at 11 53 45" src="https://github.com/Automattic/wp-calypso/assets/71137829/fbc90df3-edcb-4a60-8770-f9fedaae0cc0">

jjchrisdiehl commented:
**Promotional period updates:**

I think we could trim down the promotional period text since it directly follows the TermsOfService (TosText) which covers similar ground:

> You agree to our [Terms of Service](https://wordpress.com/tos/) and authorize your payment method to be charged on a recurring basis until you cancel, which you can do at any time. You understand [how your subscription works](https://wordpress.com/support/manage-purchases/automatic-renewal/) and [how to cancel](https://wordpress.com/support/manage-purchases/cancel-a-purchase/). 

> The promotional period for your Professional Email subscription for chrisandhandsometakeabreak.blog lasts from Nov 15, 2023 to Feb 15, 2024. On Feb 15, 2024 we will attempt to renew your subscription at the reduced price of C$35.93. All subsequent renewals will be charged for the regular subscription price of C$48. ~You will receive at least one email notice 14 days before being billed, and can [update your payment method](https://wordpress.com/support/payment/#edit-payment-details) or [manage your subscription](http://calypso.localhost:3000/purchases/subscriptions/chrisandhandsometakeabreak.wordpress.com) at any time.~

We could possibly change or reduce the following links:

- You understand [how your subscription works] -> https://wordpress.com/support/manage-purchases/automatic-renewal/ 
- and [how to cancel] -> https://wordpress.com/support/manage-purchases/cancel-a-purchase/
- and can [update your payment method] -> https://wordpress.com/support/payment/#edit-payment-details
- or [manage your subscription] -> https://wordpress.com/purchases/subscriptions/[YOUR_SITE]

Since the first two links go to the same support page, couldn't they be consolidated? 

As for the second two links, I'm not sure these are necessarily required. The VISA requirement from 2020 says `Link or other simple mechanism to enable the cardholder to easily cancel any subsequent transactions online.` 

Since we already link to the 'how to cancel' support page, could that cover this requirement? Or, does the link need to go directly to their purchase management page (which is also indirect, since the purchase in the cart doesn't even have an individual manage purchase page yet).

As for the email notice, 

> Merchants must also send a reminder notification (i.e., email or SMS/text) including a link to online cancellation policy at least seven (7) days before initiating a recurring transaction if:
A trial period, introductory offer, or promotional period is about to expire.
The nature of the recurring agreement has changed (for example, the price or billing period).

So there isn't a requirement to explain this to them, just a requirement to email them on time with a notice of the recurring transaction. I think we could remove this from the terms of service line item.

Commit:
Reorganize tos terms

jjchrisdiehl commented:
> Hello, this is my bad, but could we replace the icon on the refund line to use the TRASNFORM icon?
> 
> <img alt="Screenshot 2023-11-16 at 11 53 45" width="122" src="https://user-images.githubusercontent.com/71137829/283424807-fbc90df3-edcb-4a60-8770-f9fedaae0cc0.png">

Hey @JanaMW27 are you referring to the credit card icon next to the money back guarantee line?
![image](https://github.com/Automattic/wp-calypso/assets/16580129/1d0ca659-0846-43b6-8060-ba07a48d2ea8)

sirbrillig commented on packages/composite-checkout/src/components/checkout-steps.tsx:
```diff
@@ -13,6 +13,8 @@ import {
 	useMemo,
 } from 'react';
+/* eslint-disable no-restricted-imports */
+import isJetpackCheckout from 'calypso/lib/jetpack/is-jetpack-checkout';
```
Unfortunately we can't import things from the `client` directory into things in the `packages` directory; only the other way around. If this npm package is loaded by some other app on the internet, it won't have access to this function.

Also, the `composite-checkout` package doesn't know anything about wpcom or Jetpack. It's a general-purpose form and transaction system that could be used for anything, so we'll have to figure out another way to change these styles for this PR.

jjchrisdiehl replied:
I'll remove this since it really isn't needed.

sirbrillig commented on packages/wpcom-checkout/src/checkout-version-checker.ts:
```diff
@@ -19,3 +19,16 @@ export function hasCheckoutVersion( version: string ) {
+ * @returns - string | null
+ */
+
+export function getQueryParam( param: string ) {
```
This function isn't used in this PR so we can remove it for now.

jjchrisdiehl replied:
Removing this one too.

sirbrillig commented on config/development.json:
```diff
@@ -39,6 +39,7 @@
 		"checkout/google-pay": true,
 		"checkout/vat-form": true,
 		"checkout/checkout-version": true,
+		"checkout/tos-foldable-update": true,
```
These config flags aren't used in this PR, so we can remove them.

sirbrillig commented on client/my-sites/checkout/src/components/checkout-terms.tsx:
```diff
@@ -56,18 +77,47 @@ export default function CheckoutTerms( { cart }: { cart: ResponseCart } ) {
+						<FoldableCard
+							compact
+							className="checkout__terms-foldable-card"
+							header="Read more"
```
We should translate this text, here and in the screen reader prop.

sirbrillig commented on client/my-sites/checkout/src/components/checkout-terms.tsx:
```diff
@@ -56,18 +77,47 @@ export default function CheckoutTerms( { cart }: { cart: ResponseCart } ) {
+					{ ! isGiftPurchase && <TitanTermsOfService cart={ cart } /> }
+
+					<TermsCollapsedContent>
+						<FoldableCard
```
The first time I saw this button, I tried clicking on it but nothing happened. I guess you have to click on the little arrow next to it. This is not a blocker but I wonder if there's a way to make the text itself clickable without too much work? It looks like `FoldableCard` might have an `onClick` property, but I'm not sure if that is what we need... 🤔

jjchrisdiehl replied:
Resolved this with `isClickableHeader` prop 👍

sirbrillig commented on client/my-sites/checkout/src/components/payment-method-step.tsx:
```diff
@@ -60,15 +94,21 @@ export default function BeforeSubmitCheckoutHeader() {
 			{ ! hasCheckoutVersion( '2' ) && (
 				<WPOrderReviewSection>
-					<NonTotalPrices>
+					<NonTotalPrices showToSFoldableCard={ showToSFoldableCard }>
```
I'm not sure if it's here or somewhere else, but there's something weird about the spacing when this feature is disabled.

Before             |  After
:-------------------------:|:-------------------------:
<img width="600" alt="Screenshot 2023-11-16 at 3 52 36 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/1b79dacc-d61f-440f-8fd7-c52a1053f1e4"> | <img width="616" alt="Screenshot 2023-11-16 at 3 50 55 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/6a5a718d-51a3-48e6-a5ac-1e4a02ebd735">

jjchrisdiehl replied:
This is now fixed!

sirbrillig commented on client/my-sites/checkout/src/components/checkout-terms.tsx:
```diff
@@ -56,18 +77,47 @@ export default function CheckoutTerms( { cart }: { cart: ResponseCart } ) {
+							header="Read more"
+							screenReaderText="Read more"
+						>
+							{ shouldShowRefundPolicy && <RefundPolicies cart={ cart } /> }
```
Is there a case where none of these is visible? I know we're not going to launch this test for Jetpack right now but here's what I see right now on a Jetpack checkout view when I click to expand the card:

<img width="666" alt="Screenshot 2023-11-16 at 3 58 30 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/a198bc25-c8a3-41a9-87c3-549daf6ee742">

I think we should hide the "Read more" button entirely if that's the case.

jjchrisdiehl replied:
This has been updated to default to the International Fee notice so at least something is showing in the foldable card.

sirbrillig commented on client/my-sites/checkout/src/components/CheckoutMoneyBackGuarantee.jsx:
```diff
@@ -0,0 +1,46 @@
+export function CheckoutMoneyBackGuarantee( { cart } ) {
+	return (
+		<CheckoutMoneyBackGuaranteeWrapper>
+			<StyledMaterialIcon icon="credit_card" />
```
I believe that this icon is what Jana was talking about in her comment.

sirbrillig replied:
I think I found it: https://wordpress.github.io/gutenberg/?path=/docs/icons-icon--docs 

Search for `reusableBlock`.

jjchrisdiehl replied:
All set!

sirbrillig reviewed (commented):
Largely this looks good!

Commit:
Translate tos header and set int fee as default for foldablecard

Commit:
Fix prop for conditional padding

Commit:
Remove isJetpackCheckout from composite-checkout

Commit:
Remove getQueryParam function

Commit:
Remove feature flags

sirbrillig commented on config/stage.json:
```diff
@@ -27,6 +27,7 @@
 		"checkout/google-pay": true,
 		"checkout/vat-form": true,
 		"checkout/checkout-version": true,
+		"checkout/tos-foldable-update": true,
```
I think you can remove this since it's not used in this PR.

Commit:
Update moneyback icon to reusableBlock

Commit:
Remove flag from staging config

sirbrillig reviewed (approved):
Looks good! It doesn't appear to change anything without the flag (as designed) and with the flag I see the updated content:

<img width="654" alt="Screenshot 2023-11-17 at 4 46 53 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/77cf7c37-603c-40d7-86e8-29b814842915">

jjchrisdiehl commented:
I noticed two things while testing this PR and creating the experiment:

1) The money back guarantee does not show for carts that only have a domain product within them
![image](https://github.com/Automattic/wp-calypso/assets/16580129/3d099172-dd3d-45f8-a280-354aa9051c49)

2) This treatment differs a bit too much from the control to be a good experiment. The treatment has 3 separate variations to the footer of the checkout area that all could impact the click through rate.

- The subtotal section has been moved above the Terms of Service
- The ToS is now bifurcated with the introduction of a foldable card
- We added a money back guarantee just below the submit button

We could probably still continue with the experiment, but I just wanted to highlight these potential conflicts. As for the subtotal changing positions, I'm not sure this really needs to be done in this PR. I think I moved it because I was trying to more closely match the Checkout figma, and since in the original redesign we move the subtotals to the sidebar anyway. For this PR I think it may be useful to move the subtotal back below the ToS and just measure how the foldable card and guarantee impacts conversions.

Commit:
Reorder ToS and subtotal order

Commit:
Check if cart is only domain registrations to hide guarantee

JanaMW27 commented:
Hello, can we put the return policy sentence centered for this current implementation on the current page (where the CTA takes the entire width of the space)?

For the next version, where the CTA is right aligned, we should move it to be right aligned. But for the current one, centered, it will work better to help digest the info.

sirbrillig commented on client/my-sites/checkout/src/components/checkout-terms.tsx:
```diff
@@ -56,18 +75,48 @@ export default function CheckoutTerms( { cart }: { cart: ResponseCart } ) {
+						>
+							{ shouldShowRefundPolicy && <RefundPolicies cart={ cart } /> }
+							{ ! isGiftPurchase && <BundledDomainNotice cart={ cart } /> }
+							<InternationalFeeNotice />
```
We decided earlier (https://github.com/Automattic/wp-calypso/pull/84031/files#r1396328926) that we'd remove the `shouldShowInternationalFeeNotice` condition so that there would at least be one thing always rendered inside the card.

However, I just noticed that `InternationalFeeNotice` can itself decide to be hidden, which means there's still a case where the foldable card will be empty. 😞 

To solve this I think we're going to have to refactor each of these components so that their visibility is controlled entirely from outside the component, then we can use the logic that determines that visibility to decide if the card will be empty or not, and therefore when to hide it. (Alas, there's no effective way to check if a component has rendered something.) That's not easy, since both `BundledDomainNotice` and `RefundPolicies` have a lot of possible conditions and `JetpackSocialAdvancedPricingDisclaimer` we already saw was complex.

Let's chat about this on Monday and see if we can come up with a plan so that this doesn't delay the test.

Commit:
Fix margin for checkout-term-item

Commit:
Move vsibility logic outside of FoldableCard

sirbrillig commented on client/my-sites/checkout/src/components/checkout-terms.tsx:
```diff
@@ -59,6 +63,53 @@ export default function CheckoutTerms( { cart }: { cart: ResponseCart } ) {
+	const shouldShowInternationalFeeNotice =
+		! isJetpackCheckout() && ! isJetpackNotAtomic && ! isAkismetCheckout();
+
+	const shouldShowBundledDomainNotice = () => {
```
I assume you want this to be an IIFE? If so, you need to wrap the function in parens and call it. Alternatively, you can call it when you need it below.

sirbrillig commented on client/my-sites/checkout/src/components/checkout-terms.tsx:
```diff
@@ -59,6 +63,53 @@ export default function CheckoutTerms( { cart }: { cart: ResponseCart } ) {
+		return true;
+	};
+
+	const shouldShowJetpackSocialAdvancedPricingDisclaimer = ( cart: ResponseCart ) => {
```
Same here, this is a function so it needs to be called.

Commit:
Fix callable functions in shouldRenderFoldableCard

Commit:
Refactor visibility conditions for term components

jjchrisdiehl commented:
Refactored `<RefundPolicies cart={ cart } />`, `<BundledDomainNotice cart={ cart } />`, `<InternationalFeeNotice />`, `<JetpackSocialAdvancedPricingDisclaimer />` so that the individual conditions to show or hide these 'term components' is defined within the components themselves. This allows the components to focus on just rendering their terms rather than focusing on whether they ought to be shown in any given context.

Each component has an exported function following the pattern `showComponentName()` which manages all of the conditions that need to be met to display the term in the terms of service. 

These exported functions can then be used elsewhere, such as in `checkout-terms.tsx` to conditionally render the components. An added benefit is we can also use these new functions to query if the term components _will be rendered_ before their parent component is rendered. This fixes the issue noted here https://github.com/Automattic/wp-calypso/pull/84031#discussion_r1398471496. 

An example of this can be seen here https://github.com/Automattic/wp-calypso/pull/84031/files#diff-cc6e51447d13963e6a80cc8cf38603c75159b7301de18c8cb9fc012938a369f3R65-R77

jjchrisdiehl commented:
> Hello, can we put the return policy sentence centered for this current implementation on the current page (where the CTA takes the entire width of the space)?
> 
> For the next version, where the CTA is right aligned, we should move it to be right aligned. But for the current one, centered, it will work better to help digest the info.

@JanaMW27 Could you clarify what you are looking for here? Are you referring to the sidebar plan upsell CTA? Thanks

jjchrisdiehl commented:
Looks like there is a merge conflict blocking the tests, I'll rebase this tomorrow.

JanaMW27 commented:
> @JanaMW27 Could you clarify what you are looking for here? Are you referring to the sidebar plan upsell CTA? Thanks

Aligning center the Payment CTA and the refund element:
![refund](https://github.com/Automattic/wp-calypso/assets/71137829/6414c6b0-9f10-454a-a531-05b1f29cae21)

sirbrillig commented on client/my-sites/checkout/src/components/bundled-domain-notice.jsx:
```diff
@@ -64,14 +75,18 @@ export default function BundledDomainNotice( { cart } ) {
+export default function BundledDomainNotice( { cart } ) {
 	// Hide non-translated text for non-English users.
 	// TODO: the following lines of code should be removed once all translations are ready.
 	if (
```
We can probably remove this guard while we're here. I'm sure the string is translated by now.

jjchrisdiehl commented:
The guarantee is now center aligned on desktop devices:

<img width="614" alt="image" src="https://github.com/Automattic/wp-calypso/assets/16580129/583a8f7e-50f1-4c6d-9bce-9ecc798e1d6d">

...but on mobile there is a pattern in place started by Jetpack to left align the guarantee to ensure there is space for the current implementation of the help center icon:

<img width="373" alt="image" src="https://github.com/Automattic/wp-calypso/assets/16580129/e3ecd39f-f947-4851-a99a-ce39a9cad0cd">

We can perhaps ship this as is then center align it once I rebase and fix some issues with my mobile checkout button PR (currently backlogged) https://github.com/Automattic/wp-calypso/issues/79018

sirbrillig commented on client/my-sites/checkout/src/components/bundled-domain-notice.jsx:
```diff
@@ -55,7 +55,18 @@ function getCopyForBillingTerm( cart ) {
-export default function BundledDomainNotice( { cart } ) {
+/**
+ * Use showBundledDomainNotice to manage BundleDomainNotice visibility when called.
+ * @param {*} cart
```
I assume you had to put this here to satisfy TypeScript. There's a slightly more accurate way to use types in plain JS,

```suggestion
 * @param {import('@automattic/shopping-cart').ResponseCart} cart
```

sirbrillig reviewed (commented):
One bug I found is that this PR removes the icons after the submit button for Jetpack checkout.

Before             |  This PR without flag | This PR with flag
:-------------------------:|:-------------------------:|:-------------------------:
<img width="642" alt="Screenshot 2023-11-21 at 5 34 59 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/954e3e20-8acb-43b8-a394-776e5a3b76b2"> | <img width="634" alt="Screenshot 2023-11-21 at 5 35 07 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/9df05a8d-d1d5-4f57-a50b-163c1c0c684c"> | <img width="637" alt="Screenshot 2023-11-21 at 5 35 20 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/13a26f46-9101-4b77-8c00-99c4874977d9">

Otherwise this looks good to me! I left a couple other minor comments but they are not blockers.

jjchrisdiehl commented:
> One bug I found is that this PR removes the icons after the submit button for Jetpack checkout.

Hmm I'm not seeing this on my end:

| **Calypso.live** | **calypso.localhost** |
| ----------- | ----------- |
| <img width="278" alt="image" src="https://github.com/Automattic/wp-calypso/assets/16580129/7238e2ab-25c2-46bc-af4e-1a62cb9cdf3f"> | <img width="280" alt="image" src="https://github.com/Automattic/wp-calypso/assets/16580129/bb0cdeaa-97fc-47f3-b82e-67922ddf1d24"> |

sirbrillig commented:
🤔 Interesting. Maybe it depends on the type of Jetpack product? Or I'm doing something wrong. I'll ping you on Slack.

sirbrillig commented on client/my-sites/checkout/src/components/wp-checkout.tsx:
```diff
@@ -665,7 +668,15 @@ export default function WPCheckout( {
+						submitButtonFooter={
+							// Temporarily disabling this lint rule until hasCheckoutVersion is removed
+							// eslint-disable-next-line no-nested-ternary
+							isJetpackCheckout() ? (
```
This won't work the way you think because `isJetpackCheckout()` actually means "is jetpack _siteless_ checkout". They use the wording "jetpack checkout" all over the code to refer to that, but it doesn't mean "checkout for a jetpack site" 😭 

You need another way to see if this is a jetpack site, and fortunately we already have one hidden inside `JetpackCheckoutSeals`: `hasCartJetpackProductsOnly` should work. You could even take that logic out of that component the way we did with the term components.

Commit:
Center guarantee on desktop

Commit:
Remove translation fallback guard from bundle domain notice

Commit:
Move Jetpack seals visibility logic outside of component

sirbrillig reviewed (approved):
Looks great!

Nice work refactoring the terms components.
